### PR TITLE
rtfm: init at 0.2.2

### PR DESCRIPTION
**Pull request by sund3RRR**
###### Description of changes
It's a dash/docset reader with built in documentation for Crystal and GTK APIs. It's written in [Crystal](https://crystal-lang.org/) using [GTK4 bindings](https://github.com/hugopl/gtk4.cr).
https://github.com/hugopl/rtfm

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by sund3RRR** [comment]
@drupol, @SuperSandro2000 could you please review this PR? It's been 3 days and no one wants to :(

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (drupol) [inline pkgs/applications/misc/rtfm/gtk-docs.nix:9]: Use the `finalAttrs` pattern please

Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/gtk-docs.nix:9]: I did not quite understand how `finalAttrs` work and what it is used for.

Comment (drupol) [reply pkgs/applications/misc/rtfm/gtk-docs.nix:9]: Let me share a couple of useful links regarding this:

History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293

Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes

Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/gtk-docs.nix]: Thank you so much. There are many contentious issues when it comes to packaging. At the beginnings it is not easy to understand, i hope i did it right.

Comment (drupol) [reply pkgs/applications/misc/rtfm/gtk-docs.nix]: I'll review it when I'm back on a decent laptop

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (drupol) [inline pkgs/applications/misc/rtfm/gtk-docs.nix:12]:
```suggestion
  adw1_version = "1.3.3";
```

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (drupol) [inline pkgs/applications/misc/rtfm/gtk-docs.nix:16]:
```suggestion
    rev = "refs/tags/${finalAttrs.adw1_version}";
```

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (drupol) [inline pkgs/applications/misc/rtfm/gtk-docs.nix:20]:
```suggestion
  gtk4_version = "4.11.4";
```

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (drupol) [inline pkgs/applications/misc/rtfm/gtk-docs.nix:24]:
```suggestion
    rev = "refs/tags/${finalAttrs.gtk4_version}";
```

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (drupol) [inline pkgs/applications/misc/rtfm/gtk-docs.nix:28]:
```suggestion
  pango_version = "1.50.14";
```

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (drupol) [inline pkgs/applications/misc/rtfm/gtk-docs.nix:32]:
```suggestion
    rev = "refs/tags/${finalAttrs.pango_version}";
```

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/rtfm/default.nix
Comment (drupol) [inline pkgs/applications/misc/rtfm/default.nix:38]: Could you please add a comment on why this patch is required, for each patch in this list?

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/rtfm/default.nix
Comment (drupol) [inline pkgs/applications/misc/rtfm/default.nix:59]: Could you please add the `platforms` attribute with proper values? So we know on which architecture this software is available.

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/applications/misc/rtfm/default.nix
Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/default.nix:59]: Can i set `platforms = [ "i686-linux" "x86_64-linux" ];;` because i can't test on `aarch64` and `darwin`?

File: pkgs/applications/misc/rtfm/default.nix
Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/default.nix]: Sure

File: pkgs/applications/misc/rtfm/default.nix
Comment (drupol) [reply pkgs/applications/misc/rtfm/default.nix:68]: Just set the platforms the software you're trying to add supports, then the CI will test on these architectures.

Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/default.nix:68]: I set `platforms.linux`. Theoretically, it might run on `darwin`, maybe i'll setup OpenCore later on my PC.

Comment (drupol) [reply pkgs/applications/misc/rtfm/default.nix:68]: Or you can ask to the Darwin maintainers team, they are quite reactive. Set it correctly and we'll ask them.

Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/default.nix:68]: Done. if I understand correctly how to do it
```nix
platforms = platforms.linux ++ platforms.darwin;
```

Comment (drupol) [reply pkgs/applications/misc/rtfm/default.nix:68]: Yes!

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/gtk-docs.nix:76]: This should also get a meta section

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/gtk-docs.nix:42]:
```suggestion
```
only one should be required here

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/gtk-docs.nix:39]: gobject-introspection goes to nativeBuildInputs

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/gtk-docs.nix:34]: Do we really must pin those 3?

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/gtk-docs.nix:10]: It should get pname+version even if version is just dummy or from rtfm or so

File: pkgs/applications/misc/rtfm/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/default.nix:72]:
```suggestion
```
that's already the default

File: pkgs/applications/misc/rtfm/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/default.nix:73]:
```suggestion
```
please check if buildCrystalPackage implies this

File: pkgs/applications/misc/rtfm/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/default.nix:63]:
```suggestion
    tar -xf ${crystal-docs}
    mv crystal-${crystal_doc_version}-docs/ crystal-docs/
```

File: pkgs/applications/misc/rtfm/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/default.nix:16]: Should we passthru this?

Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/default.nix:16]: Like this?
```nix
let
  passthru = {
    gtk-docs = callPackage ./gtk-docs.nix { };
  };
in
```

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/gtk-docs.nix:34]: I think so, because they are marked as dependencies on the project's github, that is, the author assumes that during the process of building the program, all the documentation is present in the system.
```md
- Offline crystal api documentation for Crystal docset generation.
- Offline Gtk api documentation for Gtk4, Gdk4, Gsk4, libAdwaita and Pango libraries for GTK4 libraries docset generation.
```

We can build a program without documentation at all, but in this case `rtfm` will be useless. It also makes no sense to set part of the documentation, since in this case the user himself will have to generate docsets from sources.

**Comment by sund3RRR** [comment]
@hugopl could you help with one point? I generate `gi` docs using [gi-docgen](https://gnome.pages.gitlab.gnome.org/gi-docgen/). The `GDK`/`GTK` docs have a `dependencies` block, that has hard-coded dependency URLs.

Is there a way to change `Browse documentation` link to local docsets?
```toml
  [dependencies."GObject-2.0"]
  name = "GObject"
  description = "The base type system library"
  docs_url = "https://docs.gtk.org/gobject/"

  [dependencies."Gio-2.0"]
  name = "Gio"
  description = "GObject Interfaces and Objects, Networking, IPC, and I/O"
  docs_url = "https://docs.gtk.org/gio/"

  [dependencies."cairo-1.0"]
  name = "cairo"
  description = "A 2D graphics library with support for multiple output devices"
  docs_url = "https://www.cairographics.org/manual/"
  ...
```
Screenshot from 2023-07-24 13-31-29

File: pkgs/applications/misc/rtfm/default.nix
Comment (SuperSandro2000) [reply pkgs/applications/misc/rtfm/default.nix]: normally we inherit into passthru so that overriding passthru is not changing things

File: pkgs/applications/misc/rtfm/gtk-docs.nix
Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/gtk-docs.nix:36]: I didn't find docs in gtk/pango/libadwaita nix packages.

**Review by drupol — APPROVED** [review]

**Comment by sund3RRR** [comment]
After force-pushing last commit, i noticed some checks were failed.
```
ofborg-eval — This PR does not cleanly list package outputs after merging.
```
```nix
 error: attribute 'devdoc' missing

       at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-7/pkgs/applications/misc/rtfm/default.nix:67:20:

           66|
           67|     for file in "${gtk4.devdoc}"/share/doc/*; do
             |                    ^
           68|       ln -s "$file" "gtk-doc/$(basename "$file")"
```

It happens because `gtk4.devdoc` output depends in `x11Support` optional argument
```nix
outputs = [ "out" "dev" ] ++ lib.optionals x11Support [ "devdoc" ];
```
So, maybe i need to override `x11Support = true`?

**Comment by SuperSandro2000** [comment]
> So, maybe i need to override `x11Support = true`?

seems like it

**Comment by sund3RRR** [comment]
> > So, maybe i need to override `x11Support = true`?
> 
> seems like it

Nothing happened, i don't know why and how to fix it

File: pkgs/applications/misc/rtfm/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/default.nix:43]:
```suggestion
    substituteInPlace Makefile \
      --replace "crystal run src/create_crystal_docset.cr" "crystal src/create_crystal_docset.cr ${crystal}/share/doc/crystal/api/" \
      --replace "crystal run src/create_gtk_docset.cr" "crystal src/create_gtk_docset.cr gtk-doc/"
```

File: pkgs/applications/misc/rtfm/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/default.nix:58]:
```suggestion
    gtk4'
```

File: pkgs/applications/misc/rtfm/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/default.nix:14]:
```suggestion
}:

let
  gtk4' = gtk4.override { x11Support = true; };
in
```

**Comment by hugopl** [comment]
> Is there a way to change `Browse documentation` link to local docsets?

I can reproduce this on Archlinux too and create [an issue](https://github.com/hugopl/rtfm/issues/6).

File: pkgs/applications/misc/rtfm/default.nix
Comment (SuperSandro2000) [inline pkgs/applications/misc/rtfm/default.nix:75]:
```suggestion
    for file in "${gtk4'.devdoc}"/share/doc/*; do
      ln -s "$file" "gtk-doc/$(basename "$file")"
    done

    for file in "${pango.devdoc}"/share/doc/*; do
      ln -s "$file" "gtk-doc/$(basename "$file")"
    done

    for file in "${libadwaita.devdoc}"/share/doc/*; do
```
Those are taken directly from the inputs, not buildInputs where you did the override

**Comment by hugopl** [comment]
@sund3RRR would you mind in creating a PR for rtfm README with the instructions to install this NixOS package?

**Comment by sund3RRR** [comment]
> @sund3RRR would you mind in creating a PR for rtfm README with the instructions to install this NixOS package?

Sure, but these instructions will be useless until this PR is approved

**Comment by hugopl** [comment]
Yeah, sure, I mean, when it gets approved/merged.

**Review by drupol — APPROVED** [review]
Thanks for your contribution, it's ok for me now.

File: pkgs/applications/misc/rtfm/default.nix
Comment (hugopl) [inline pkgs/applications/misc/rtfm/default.nix:33]: I can fix that upstream.

File: pkgs/applications/misc/rtfm/default.nix
Comment (hugopl) [inline pkgs/applications/misc/rtfm/default.nix:35]: Usually distros run these command for any packages that install files on specific dirs used by these commands.

File: pkgs/applications/misc/rtfm/patches/make.patch
Comment (hugopl) [inline pkgs/applications/misc/rtfm/patches/make.patch:9]: Why you guys don't use the gi-crystal that the project depends on? Do you guys have a gi-crystal nix package? if so, why?

File: pkgs/applications/misc/rtfm/patches/make.patch
Comment (hugopl) [inline pkgs/applications/misc/rtfm/patches/make.patch:20]: I added this change upstream at https://github.com/hugopl/rtfm/commit/8833114005f37b0ded3b62a8d296f02a0b1cfeaa

File: pkgs/applications/misc/rtfm/default.nix
Comment (hugopl) [inline pkgs/applications/misc/rtfm/default.nix:42]: Could you explain this error? why does it happen? So I can fix it upstream too.

File: pkgs/applications/misc/rtfm/patches/make.patch
Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/patches/make.patch:9]: I packed `gi-crystal` in `nixpkgs` when I packed the `collision`. You can see these PR's: https://github.com/NixOS/nixpkgs/pull/243914 
https://github.com/NixOS/nixpkgs/pull/242681
There is an explanation why we decided to do it this way.

Shortly, nix collects crystal dependencies in separated derivations, then creates a symlimks on it in lib/ directory of project. I am creating a patch when packing `gi-crystal` to make it work with symlinks. The main problem is that all derivations in nix store are read-only and changes in one derivation will change drv hash and all hashes, that depend on it. When we run `./bin/gi-crystal` it fails with error `Permission denied`, because it cannot write to read-only directory.

So, `gi-crystal` in nixpkgs works differently due to this nix features. All you need to do is add `gi-crystal` in build inputs, then run `gi-crystal`. It will scan `lib/` directory with all symlinks, then generate all bindings in `lib/gi-crystal` directory and copy self package to this.

File: pkgs/applications/misc/rtfm/default.nix
Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/default.nix:42]: It happens because `rtfm` copies docs from nix derivation, but all files in nix-store are read-only, so at first we need to add write permissions to the copied file. This error is nix-specific, but it cannot break `rtfm` on other distros i think.

File: pkgs/applications/misc/rtfm/patches/make.patch
Comment (drupol) [reply pkgs/applications/misc/rtfm/patches/make.patch:9]: > Shortly, nix collects crystal dependencies in separated derivations, then creates a symlimks on it in lib/ directory of project. I am creating a patch when packing `gi-crystal` to make it work with symlinks. The main problem is that all derivations in nix store are read-only and changes in one derivation will change drv hash and all hashes, that depend on it. When we run `./bin/gi-crystal` it fails with error `Permission denied`, because it cannot write to read-only directory.

Jumping into the convo here... How about copying the `gi-crystal` derivation into the current rtfm directory and then use it ? That would fix both issues. You will be re-using the existing `gi-cristal` and it's directory is writable since you copied it into the `rtfm` build directory.

WDYT? Is it feasible?

Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/patches/make.patch:9]: > How about copying the gi-crystal derivation into the current rtfm directory and then use it ?

This is exactly what I did when packing `gi-crystal`. But I actually automated this action so as not to do it constantly for different packages. 

I think that's better, don't you?

https://github.com/NixOS/nixpkgs/pull/243914/files#diff-1a3ec3be7dadadd498d82613b7eaba4d02f11f8470913c5cbdb7d54932c63954R53

Comment (drupol) [reply pkgs/applications/misc/rtfm/patches/make.patch:9]: What I mean is to copy the output of `gi-crystal` into `rtfm` and then use it.
Once the compilation is over, you delete it and only copy the relevant `rtfm` files in `$out`.

WDYT?

Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/patches/make.patch:9]: > What I mean is to copy the output of gi-crystal into rtfm and then use it.
> Once the compilation is over, you delete it and only copy the relevant rtfm files in $out.

I'm sorry but I don't understand what's the difference. If we want to make it work without patching `Makefile`, then we need to do several steps:
```bash
mkdir lib/gi-crystal
cp -r ${gi-crystal}/* lib/gi-crystal
mkdir bin/
mv lib/gi-crystal/bin/gi-crystal bin/
chmod +w lib/gi-crystal/src/auto
```

I think it is just unnecessary steps to be done on each package requires gi-crystal. 

And even if we do that, we still have to patch this to remove `shards install` command, because it will crash building due to `cannot fetch remote host github.com` something like this, I definitely don't remember what that error sounds like.

WDYT?

Comment (drupol) [reply pkgs/applications/misc/rtfm/patches/make.patch:9]: I don't see the issue to do that all the time. I prefer that than patching the src.

But anyway, I was just passing by here thinking to provides new ideas, but I see that you already studied the question so I won't bother. I never used gi-crystal this is the reason why I might be wrong too.

Comment (sund3RRR) [reply pkgs/applications/misc/rtfm/patches/make.patch:9]: > I don't see the issue to do that all the time. I prefer that than patching the src.
> 
> But anyway, I was just passing by here thinking to provides new ideas, but I see that you already studied the question so I won't bother. I never used gi-crystal this is the reason why I might be wrong too.

Thanks for suggesting solutions to this problem. Patching sources was the reason why I didn't want to pack `gi-crystal`. My best idea is to make `gi-crystal` portable, this fits well with the nix ideology, but changes program behaviour.

Speaking of using gi-crystal, I understand the steps involved in implementing your idea. But others may not realize that you need to copy the `gi-crystal` directory, mv `gi-crystal` binary and change the write permissions.

Comment (drupol) [reply pkgs/applications/misc/rtfm/patches/make.patch:9]: Ok... sad but ok.

It's weird that it needs to write in the `gi-crystal` directory.

Comment (hugopl) [reply pkgs/applications/misc/rtfm/patches/make.patch:9]: Looks like you guys are re-doing what shards already does, but at a package level. I mean... if derivations means a packaged code.

**Pull request merged**
